osu.Game/Database/StandardisedScoreMigrationTools.cs

```diff
@@ -274,53 +274,79 @@ public static long ConvertFromLegacyTotalScore(ScoreInfo score, LegacyBeatmapCon
+                    // Similarly, estimate the maximum magnitude of the combo portion in standardised score.
+                    // Roughly corresponds to integrating f(combo) = combo ^ COMBO_EXPONENT (omitting constants)
+                    double maximumAchievableComboPortionInStandardisedScore = Math.Pow(maximumLegacyCombo, 1 + ScoreProcessor.COMBO_EXPONENT);
+
+                    double comboPortionInScoreV1 = maximumAchievableComboPortionInScoreV1 * comboProportion / score.Accuracy;
+
+                    // This is - roughly - how much score, in the combo portion, the longest combo on this particular play would gain in score V1.
+                    double comboPortionFromLongestComboInScoreV1 = Math.Pow(score.MaxCombo, 2);
+                    // Same for standardised score.
+                    double comboPortionFromLongestComboInStandardisedScore = Math.Pow(score.MaxCombo, 1 + ScoreProcessor.COMBO_EXPONENT);
+
+                    // Calculate how many times the longest combo the user has achieved in the play can repeat
+                    // without exceeding the combo portion in score V1 as achieved by the player.
+                    // This is a pessimistic estimate; it intentionally does not operate on object count and uses only score instead.
+                    double maximumOccurrencesOfLongestCombo = Math.Floor(comboPortionInScoreV1 / comboPortionFromLongestComboInScoreV1);
+                    double comboPortionFromRepeatedLongestCombosInScoreV1 = maximumOccurrencesOfLongestCombo * comboPortionFromLongestComboInScoreV1;
+
+                    double remainingComboPortionInScoreV1 = comboPortionInScoreV1 - comboPortionFromRepeatedLongestCombosInScoreV1;
+                    // `remainingComboPortionInScoreV1` is in the "score ballpark" realm, which means it's proportional to combo squared.
+                    // To convert that back to a raw combo length, we need to take the square root...
+                    double remainingCombo = Math.Sqrt(remainingComboPortionInScoreV1);
+                    // ...and then based on that raw combo length, we calculate how much this last combo is worth in standardised score.
+                    double remainingComboPortionInStandardisedScore = Math.Pow(remainingCombo, 1 + ScoreProcessor.COMBO_EXPONENT);
+
+                    double lowerEstimateOfComboPortionInStandardisedScore
+                        = maximumOccurrencesOfLongestCombo * comboPortionFromLongestComboInStandardisedScore
+                          + remainingComboPortionInStandardisedScore;
+
+                    // Compute approximate upper estimate new score for that play.
+                    // This time, divide the remaining combo among remaining objects equally to achieve longest possible combo lengths.
+                    // There is no rigorous proof that doing this will yield a correct upper bound, but it seems to work out in practice.
```

Comment:
I'm good with your changes.
My only comment would be that I can provide a proof for this if you really want.

Reply:
That would be nice if you have one, just for my peace of mind.